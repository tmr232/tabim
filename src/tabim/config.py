import enum

import attr


class LyricsPosition(str, enum.Enum):
    Top = "top"
    Bottom = "bottom"


@attr.s(auto_attribs=True, slots=True)
class HeaderConfig:
    show_title: bool = True
    center_title: bool = True
    show_subtitle: bool = True
    show_artist: bool = True
    show_album: bool = True
    show_music: bool = True
    show_words: bool = True
    show_tab: bool = True
    show_copyright: bool = True


@attr.s(auto_attribs=True, slots=True)
class LineConfig:
    line_length: int = 60
    show_lyrics: bool = True
    show_bar_numbers: bool = True
    lyrics_position: LyricsPosition = LyricsPosition.Top
    split_sections: bool = True
    show_cont: bool = True


@attr.s(auto_attribs=True, slots=True)
class RenderConfig:
    header: HeaderConfig = attr.Factory(HeaderConfig)
    line: LineConfig = attr.Factory(LineConfig)
