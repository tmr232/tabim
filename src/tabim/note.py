from __future__ import annotations

import io
from functools import reduce
from itertools import repeat, chain, groupby
from operator import attrgetter, add
from typing import Sequence, Optional, List

import attr
import guitarpro
from more_itertools import windowed, interleave


@attr.s(auto_attribs=True)
class AsciiNote:
    note: str = ""
    start: int = 0

    @property
    def head(self):
        return self.note[: self.start]

    @property
    def tail(self):
        return self.note[self.start :]


@attr.s(auto_attribs=True)
class TabBeat:
    notes: Sequence[Optional[guitarpro.Note]]  #: The notes in the beat

    def render(self, width: int, prev_beat: Optional[TabBeat]) -> Sequence[str]:
        if prev_beat:
            prev_notes = prev_beat.notes
        else:
            prev_notes = repeat(None)

        ascii_notes = [
            render_note(note, prev) for note, prev in zip(self.notes, prev_notes)
        ]

        # Get the minimal drawing width
        max_head = max(len(note.head) for note in ascii_notes)
        max_tail = max(len(note.tail) for note in ascii_notes)
        min_width = max_head + max_tail
        # Set the draw width based on minimal and desired
        draw_width = max(min_width, width)
        draw_tail = draw_width - max_head

        drawn_notes = []
        for note in ascii_notes:
            drawn_notes.append(
                "-" * (max_head - len(note.head) + 1)
                + note.note
                + "-" * (draw_tail - len(note.tail))
            )

        return drawn_notes


def render_note(
    note: Optional[guitarpro.Note],
    prev: Optional[guitarpro.Note],
) -> AsciiNote:
    if not note:
        return AsciiNote()
    if note.type == guitarpro.NoteType.normal:
        if prev and prev.effect.hammer:
            if prev.value > note.value:
                return AsciiNote(f"p{note.value}", 1)
            else:
                return AsciiNote(f"h{note.value}", 1)

        if note.effect.isHarmonic:
            if isinstance(note.effect.harmonic, guitarpro.NaturalHarmonic):
                return AsciiNote(f"<{note.value}>", 1)

        if prev and prev.effect.slides:
            if prev.effect.slides[0] == guitarpro.SlideType.legatoSlideTo:
                if prev.value > note.value:
                    return AsciiNote(f"\\{note.value}", 1)
                else:
                    return AsciiNote(f"/{note.value}", 1)
            if prev.effect.slides[0] == guitarpro.SlideType.shiftSlideTo:
                return AsciiNote(f"s{note.value}", 1)

        if note.effect.isBend:
            bend = note.effect.bend
            if bend.type in (guitarpro.BendType.bend, guitarpro.BendType.bendRelease):
                parts = [f"{note.value}"]
                for frm, to in windowed(
                    chain([0], map(attrgetter("value"), bend.points)), 2
                ):
                    if frm < to:
                        parts.append(f"b{note.value + to // 2}")
                    elif frm > to:
                        parts.append(f"r{note.value + to // 2}")
                return AsciiNote("".join(parts), 0)

        if note.effect.isTrill:
            return AsciiNote(f"{note.value}tr{note.effect.trill.fret}")

        if note.effect.vibrato:
            return AsciiNote(f"{note.value}~")

        if note.effect.isDefault or note.effect.hammer or note.effect.slides:
            # Hammers affect the _next_ note.
            # Same goes for slides
            return AsciiNote(f"{note.value}")

    if note.type == guitarpro.NoteType.tie:
        if prev:
            return AsciiNote(f"({prev.value})", 1)

    if note.type == guitarpro.NoteType.rest:
        return AsciiNote()

    if note.type == guitarpro.NoteType.dead:
        return AsciiNote("x")

    raise NotImplementedError()


@attr.s(auto_attribs=True)
class TabMeasure:
    beats: Sequence[TabBeat]
    divisions: Sequence[int]
    strings: int


def get_end_beat(tab_measure: TabMeasure) -> TabBeat:
    notes = [None for _ in range(tab_measure.strings)]
    for beat in tab_measure.beats:
        for i, note in enumerate(beat.notes):
            if note:
                notes[i] = note

    return TabBeat(notes=notes)


def parse_measure(measure: guitarpro.Measure, strings: int = 6):
    notes = []
    timestamps = set()
    for voice in measure.voices:
        for beat in voice.beats:
            timestamps.add(beat.startInMeasure)
            for note in beat.notes:
                notes.append(note)

    beats = []
    for _, grouped_notes in groupby(
        sorted(notes, key=attrgetter("beat.startInMeasure")),
        key=attrgetter("beat.startInMeasure"),
    ):
        beat_notes: List[Optional[guitarpro.Note]] = [None for _ in range(strings)]
        for note in grouped_notes:
            beat_notes[note.string - 1] = note
        beats.append(TabBeat(notes=beat_notes))

    divisions = [
        measure.length // (end - start)
        for start, end in windowed(chain(sorted(timestamps), [measure.length]), 2)
    ]

    return TabMeasure(beats=beats, divisions=divisions, strings=strings)


def draw_measure(
    tab_measure: TabMeasure,
    quarter_width: int = 8,
    prev_beat: Optional[TabBeat] = None,
) -> str:
    rendered_beats = []
    for (prev, beat), division in zip(
        windowed(chain([prev_beat], tab_measure.beats), 2), tab_measure.divisions
    ):
        width = min(quarter_width, quarter_width * 4 // division)
        rendered_beats.append(beat.render(width=width, prev_beat=prev))

    output = io.StringIO()
    for line in zip(*rendered_beats):
        print("".join(line), file=output)

    return output.getvalue()


def _merge(column_a: str, column_b: str) -> str:
    return "\n".join(map(add, column_a.splitlines(), column_b.splitlines()))


def render_line(line_measures: Sequence[str]) -> str:
    tuning = "\n".join(reversed("EADGBe"))
    border = "\n".join(("|" * 6))
    parts = chain([tuning, border], interleave(line_measures, repeat(border)))
    return reduce(_merge, parts)


def render_track(
    track: guitarpro.Track,
    strings: int = 6,
    quarter_width: int = 8,
    line_length: int = 79,
) -> str:
    rendered_measures = []
    prev_beat = None
    for measure in track.measures:
        tab_measure = parse_measure(measure, strings=strings)
        rendered_measures.append(
            draw_measure(tab_measure, quarter_width=quarter_width, prev_beat=prev_beat)
        )
        prev_beat = get_end_beat(tab_measure)

    cur_length = 0
    lines = [[]]
    for rendered_measure in rendered_measures:
        measure_length = len(rendered_measure.splitlines()[0])
        if cur_length + measure_length < line_length:
            lines[-1].append(rendered_measure)
            cur_length += measure_length
        else:
            lines.append([rendered_measure])
            cur_length = measure_length

    rendered_lines = []
    bar_number = 1
    for line in lines:
        rendered_lines.append((bar_number, render_line(line)))
        bar_number += len(line)

    output = io.StringIO()

    for bar_number, line in rendered_lines:
        print(bar_number, file=output)
        print(line, file=output)
        print(file=output)

    return output.getvalue()
