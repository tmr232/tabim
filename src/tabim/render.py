import functools
import operator
from typing import Optional, NewType, List

import attr


@attr.s(auto_attribs=True)
class DisplayNote:
    cont_in: bool = False
    cont_out: bool = False
    cont: bool = False
    fret: Optional[int] = None


@attr.define
class RenderedNote:
    prefix: str
    postfix: str
    pre_cont: str
    post_cont: str

    def __str__(self):
        return f"{self.pre_cont}_{self.prefix}.{self.postfix}_{self.post_cont}"


@attr.s(auto_attribs=True)
class Column:
    notes: List[DisplayNote]
    division: int


def render_note(note: DisplayNote) -> RenderedNote:
    if note.cont_in:
        pre_cont = "="
    else:
        pre_cont = "-"
    if note.cont_out:
        post_cont = "="
    else:
        post_cont = "-"
    prefix = ""
    if note.cont:
        postfix = "="
    elif note.fret is None:
        postfix = "-"
    else:
        postfix = str(note.fret)

    return RenderedNote(
        prefix=prefix,
        postfix=postfix,
        pre_cont=pre_cont,
        post_cont=post_cont,
    )


def concat_columns(col1: str, col2: str) -> str:
    return "\n".join(map(operator.add, col1.splitlines(), col2.splitlines()))


import rich


def render_column(column: Column, quarter_width: int) -> str:
    rendered: List[RenderedNote] = list(map(render_note, column.notes))
    max_prefix = max(len(n.prefix) for n in rendered)
    max_postfix = max(len(n.postfix) for n in rendered)

    target_width = quarter_width * 4 // column.division

    # max_prefix = max(1, max_prefix)
    max_postfix = max(target_width, max_postfix)

    return "\n".join(
        note.prefix.rjust(max_prefix, note.pre_cont)
        + note.postfix.ljust(max_postfix, note.post_cont)
        for note in rendered
    )


def render_columns(columns, quarter_width):
    return functools.reduce(
        concat_columns,
        map(lambda col: render_column(col, quarter_width), columns),
        ("|-\n" * 6),
    )
