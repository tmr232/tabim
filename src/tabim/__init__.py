__all__ = ["note", "measure", "render"]

import tabim.note
