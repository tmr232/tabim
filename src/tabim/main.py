import io

import typer
from pathlib import Path
from typing import Optional
import guitarpro

from tabim.measure import build_measure, parse_notes
from tabim.render import render_columns


def main(gp_path: Path, out_path: Optional[Path] = None):
    with gp_path.open("rb") as stream:
        tab = guitarpro.parse(stream)

    track = tab.tracks[0]

    output = io.StringIO()

    for bar, gp_measure in enumerate(track.measures, start=1):

        measure = build_measure(gp_measure)
        print(bar, file=output)
        print(render_columns(measure.columns, 8), file=output)

    if out_path:
        with out_path.open("w") as f:
            f.write(output.getvalue())
    else:
        print(output.getvalue())


if __name__ == "__main__":
    typer.run(main)
