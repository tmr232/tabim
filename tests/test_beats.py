from __future__ import annotations

import enum
import io
import re
from collections import deque
from functools import reduce
from itertools import groupby, chain, cycle, repeat
from operator import attrgetter
from typing import Iterable, Any, Optional, Iterator, Sequence, List

import attr
import guitarpro
import pytest
import rich
from more_itertools import windowed, chunked, interleave

from tabim.note import render_note
from tabim.render import concat_columns
from tests.conftest import get_sample


def unnest(iterable: Iterable[Any], *attrs: str, extra: Optional[str] = None):
    items = deque(iterable)

    if attrs:
        sentinel = object()
        items.append(sentinel)
        for getter in map(attrgetter, attrs):
            while items:
                item = items.popleft()
                if item is sentinel:
                    break
                items.extend(getter(item))

    if extra:
        yield from map(attrgetter(extra), items)
    else:
        yield from items


@attr.s(auto_attribs=True)
class TabNote:
    note: guitarpro.Note
    prev_note: Optional[TabNote]
    is_play: bool = False
    is_cont: bool = False
    tie_note: Optional[TabNote] = None

    @property
    def fret(self):
        cur = self
        while cur.tie_note:
            cur = cur.tie_note
        return cur.note.value

    @property
    def is_tie(self):
        return bool(self.tie_note)

    @staticmethod
    def play(
        note: guitarpro.Note, prev_note: Optional[guitarpro.Note] = None
    ) -> TabNote:
        return TabNote(
            note=note,
            is_play=True,
            prev_note=prev_note,
        )

    @staticmethod
    def cont(
        note: guitarpro.Note, prev_note: Optional[guitarpro.Note] = None
    ) -> TabNote:
        return TabNote(
            note=note,
            is_cont=True,
            prev_note=prev_note,
        )

    @staticmethod
    def tie(
        note: guitarpro.Note,
        tie_note: TabNote,
    ) -> TabNote:
        # Propagate cont
        return TabNote(
            note=note,
            tie_note=tie_note,
            is_cont=tie_note.is_cont,
            prev_note=tie_note,
        )

    def set_cont(self):
        self.is_cont = True

        # Propagate up tie-notes!
        tie_note = self.tie_note
        while tie_note:
            tie_note.is_cont = True
            tie_note = tie_note.tie_note


@attr.s(auto_attribs=True)
class TabBeat:
    start: int
    notes: list[TabNote] = attr.Factory(list)
    is_measure_break: bool = False
    is_rest: bool = False
    lyric: str = ""

    @staticmethod
    def from_notes(
        start: int, notes: list[TabNote], lyric: Optional[str] = None
    ) -> TabBeat:
        return TabBeat(notes=notes, lyric=lyric, start=start)

    @staticmethod
    def measure(start: int) -> TabBeat:
        return TabBeat(is_measure_break=True, start=start)


def get_measure_beats(measure: guitarpro.Measure) -> list[guitarpro.Beat]:
    beats = unnest(measure.voices, "beats")
    return sorted(beats, key=attrgetter("start"))


def get_grouped_beats(
    measure: guitarpro.Measure,
) -> Iterator[tuple[int, Iterator[guitarpro.Beat]]]:
    return groupby(get_measure_beats(measure), key=attrgetter("start"))  # type:ignore


def parse_lyrics(
    lyric_line: guitarpro.LyricLine, track: guitarpro.Track
) -> dict[int, str]:
    timestamps = []
    for measure in track.measures[lyric_line.startingMeasure - 1 :]:
        for beat in measure.voices[0].beats:
            if beat.status != guitarpro.BeatStatus.normal:
                continue
            timestamps.append(beat.start)

    lyric_fragments = [
        "".join(parts).strip()
        for parts in chunked(re.split("([- \n]+)", lyric_line.lyrics), 2)
    ]

    return dict(zip(timestamps, lyric_fragments))


def parse_song(song: guitarpro.Song, track_number: int = 0):
    """
    The idea is that we break the song up into yet another new kind of beat.
    This itme, the beat holds notes of the following types:

        * ``Play(note)`` - this is where the note starts
        * ``Cont(note)`` - The note is still active
        * ``Tie(note, into)`` - Ties to the previous note. Also holds the original note.

    So that later we can iterate and know what to render.
    Additionally, a beat can also have a "measure-bound" type so that
    we just know to draw it.

    The idea is that as we traverse the notes, we build the beats to
    hold all the relevant information.
    Later, when we actually draw - we traverse in pairs (or triplets?) to
    know how to apply effects and continuations.
    But the basic knowledge of whether to apply continuations is already present.

    Beats will also hold info on whether they are "Strong" beats (quarters in 4/4, for example)
    and the relevant lyric fragment.
    """
    track = song.tracks[track_number]

    lyric_timestamps = parse_lyrics(song.lyrics.lines[0], track)

    tab_beats = []
    live_notes: list[Optional[TabNote]] = [None for _ in track.strings]
    for measure in track.measures:
        for timestamp, beats in get_grouped_beats(measure):
            tie_live_notes = live_notes[:]
            # Remove all ended live-notes
            for string, note in enumerate(live_notes):
                if not note:
                    continue
                if note.note.beat.start + note.note.beat.duration.time <= timestamp:
                    live_notes[string] = None

            # Collect new notes from current beats
            notes: list[Optional[TabNote]] = [None for _ in track.strings]
            new_live_notes = live_notes[:]
            tie_notes = []
            has_play = False  # Denotes whether any play-note was present in the beat
            for beat in beats:
                for note in beat.notes:
                    if note.type == guitarpro.NoteType.tie:
                        new_note = TabNote.tie(
                            note, tie_note=tie_live_notes[note.string - 1]
                        )
                        tie_notes.append(new_note)
                    else:
                        new_note = TabNote.play(
                            note, prev_note=tie_live_notes[note.string - 1]
                        )
                        has_play = True
                    new_live_notes[note.string - 1] = new_note
                    notes[note.string - 1] = new_note
            if has_play:
                for tie_note in tie_notes:
                    tie_note.set_cont()

            # Mark continuation for live notes
            for string, (note, live_note) in enumerate(zip(notes, live_notes)):
                if note:
                    continue
                if live_note and has_play:
                    notes[string] = TabNote.cont(
                        live_note.note, tie_live_notes[live_note.note.string - 1]
                    )
                    # Propagate cont
                    live_note.set_cont()

            # Update live notes
            live_notes = new_live_notes

            tab_beats.append(
                TabBeat.from_notes(
                    notes=notes,
                    lyric=lyric_timestamps.get(timestamp, ""),
                    start=timestamp,
                )
            )

        tab_beats.append(TabBeat.measure(start=measure.end))

    return tab_beats


def try_getattr(obj: Optional[Any], attr: str) -> Any:
    attrs = str.split(attr, ".")
    for name in attrs:
        if obj is None:
            break
        obj = getattr(obj, name)
    return obj


@attr.s(auto_attribs=True, slots=True)
class AsciiMeasure:
    lyrics: Sequence[str]
    strings: Sequence[Sequence[str]]

    @property
    def width(self):
        return len("".join(self.lyrics))

    def __len__(self):
        return self.width


def less_naive_render_beats(
    beats: Sequence[TabBeat], n_strings: int = 6
) -> Sequence[AsciiMeasure]:
    lyrics = []
    strings = [[] for _ in range(n_strings)]

    measures: list[AsciiMeasure] = []

    measure_break_notes = [True for _ in range(n_strings)]

    for beat in beats:
        if beat.is_measure_break:
            measures.append(AsciiMeasure(lyrics=lyrics, strings=strings))
            lyrics = []
            strings = [[] for _ in range(n_strings)]
            measure_break_notes = [True for _ in range(n_strings)]
            continue

        ascii_notes = [
            render_note(
                note=try_getattr(note, "note"),
                prev=try_getattr(note, "prev_note.note"),
            )
            for note in beat.notes
        ]

        max_head = max(len(note.head) for note in ascii_notes)
        max_tail = max(
            len(beat.lyric),
            max(len(note.tail) for note in ascii_notes),
        )

        draw_width = max(3, max_head + max_tail + 1)
        draw_tail = draw_width - max_head

        lyrics.append(" " * max_head + beat.lyric.ljust(draw_tail))
        for i, (note, ascii_note) in enumerate(zip(beat.notes, ascii_notes)):
            # No note, so we just draw the empty state
            if not note:
                strings[i].append("-" * draw_width)
                continue

            # If the previous note is a cont, we need to pad with a cont
            head_pad = "-"
            if note.prev_note and note.prev_note.is_cont:
                head_pad = "="

            # If this is a cont, pad with a cont
            tail_pad = "-"
            if note.is_cont:
                tail_pad = "="

            base = ""
            head = 0
            tail = 0
            if note.is_play or (note.is_tie and measure_break_notes[i]):
                base = ascii_note.note
                head = len(ascii_note.head)
                tail = len(ascii_note.tail)

            rendered = (
                head_pad * (max_head - head) + base + tail_pad * (draw_tail - tail)
            )

            measure_break_notes[i] = False

            strings[i].append(rendered)

    return measures


def naive_render_beats(beats: list[TabBeat], n_strings: int = 6) -> str:
    lyrics = []
    strings = [[] for _ in range(n_strings)]

    measures = []

    measure_break_notes = [True for _ in range(n_strings)]

    for beat in beats:
        if beat.is_measure_break:
            measures.append((lyrics, strings))
            lyrics = []
            strings = [[] for _ in range(n_strings)]
            measure_break_notes = [True for _ in range(n_strings)]
            continue
        width = max(3, len(beat.lyric) + 1)
        lyrics.append(beat.lyric.ljust(width))
        for i, note in enumerate(beat.notes):
            if not note:
                strings[i].append("-" * width)
                continue
            base = ""
            pad = "-"
            if note.is_play:
                base = str(note.fret)
            if note.is_cont:
                pad = "="
            if note.is_tie and measure_break_notes[i]:
                base = f"({note.fret})"

            measure_break_notes[i] = False

            strings[i].append(base.ljust(width, pad))

    output = io.StringIO()

    for i, (lyrics, strings) in enumerate(measures, start=1):
        print(i, file=output)
        print("".join(lyrics).rstrip(), file=output)
        for string in strings:
            print("".join(string).rstrip(), file=output)
        print(file=output)

    return output.getvalue()


class LyricsPosition(enum.Enum):
    Top = enum.auto()
    Bottom = enum.auto()


def render_measure(
    measure: AsciiMeasure,
    show_lyrics: bool = True,
    lyrics_position: LyricsPosition = LyricsPosition.Top,
) -> str:
    output = io.StringIO()

    def add_lyrics():
        if show_lyrics:
            print("".join(measure.lyrics), file=output)

    if lyrics_position == LyricsPosition.Top:
        add_lyrics()

    for string in measure.strings:
        print("".join(string), file=output)

    if lyrics_position == LyricsPosition.Bottom:
        add_lyrics()

    return output.getvalue()


def render_line(
    line: Sequence[AsciiMeasure],
    tuning: Sequence[str],
    show_lyrics: bool = True,
    lyrics_position: LyricsPosition = LyricsPosition.Top,
    n_string: int = 6,
) -> str:
    rendered_measures = [
        render_measure(
            measure, show_lyrics=show_lyrics, lyrics_position=lyrics_position
        )
        for measure in line
    ]

    if show_lyrics:
        if lyrics_position == LyricsPosition.Top:
            tuning_header = "\n".join([" "] + list(tuning))
            measure_separator = "\n".join(" " + "|" * n_string)
        else:  # lyrics_position == LyricsPosition.Bottom:
            tuning_header = "\n".join(list(tuning) + [" "])
            measure_separator = "\n".join("|" * n_string + " ")

    else:
        tuning_header = "\n".join(list(tuning))
        measure_separator = "\n".join("|" * n_string)

    return reduce(
        concat_columns,
        chain(
            [tuning_header],
            interleave(repeat(measure_separator), rendered_measures),
            [measure_separator],
        ),
    )


def render_measures(
    measures: Sequence[AsciiMeasure],
    line_length: int = 90,
    tuning: Sequence[str] = "EADGBe"[::-1],
    show_lyrics: bool = True,
    bar_numbers: bool = True,
    lyrics_position: LyricsPosition = LyricsPosition.Top,
) -> str:

    lines = []
    current_line = []
    current_line_length = 0
    for measure in measures:
        current_line.append(measure)
        current_line_length += len(measure)
        if current_line_length > line_length:
            lines.append(current_line)
            current_line = []
            current_line_length = 0

    lines.append(current_line)

    current_bar = 1
    output = io.StringIO()
    for line in lines:
        rendered_line = render_line(
            line,
            show_lyrics=show_lyrics,
            tuning=tuning,
            lyrics_position=lyrics_position,
        )

        if bar_numbers:
            print(current_bar, file=output)

            if lyrics_position == LyricsPosition.Bottom or not show_lyrics:
                print(file=output)

        print(rendered_line, file=output)
        print(file=output)

        current_bar += len(line)

    return strip_trailing_whitespace(output.getvalue())


def strip_trailing_whitespace(text: str):
    return "\n".join(line.rstrip() for line in text.splitlines())


def naive_render_measures(measures: Sequence[AsciiMeasure]):
    output = io.StringIO()

    for i, measure in enumerate(measures, start=1):
        print(i, file=output)
        print(render_measure(measure, show_lyrics=True), file=output)

    return strip_trailing_whitespace(output.getvalue())


@attr.s(auto_attribs=True, slots=True)
class HeaderConfig:
    show_title: bool = True
    center_title: bool = True
    show_subtitle: bool = True
    title_box: bool = True
    show_artist: bool = True
    show_album: bool = True
    show_music: bool = True
    show_words: bool = True
    show_tab: bool = True
    show_copyright: bool = True


@attr.s(auto_attribs=True, slots=True)
class LineConfig:
    line_length: int = 60
    show_lyrics: bool = True
    show_bar_numbers: bool = True
    lyrics_position: LyricsPosition = LyricsPosition.Top


@attr.s(auto_attribs=True, slots=True)
class RenderConfig:
    header: HeaderConfig = attr.Factory(HeaderConfig)
    line: LineConfig = attr.Factory(LineConfig)


def get_tuning(strings: Sequence[guitarpro.GuitarString]):
    tuning = []
    notes = "C C# D D# E F F# G G# A A# B".split()
    for string in strings:
        octave, semitone = divmod(string.value, 12)
        tuning.append(notes[semitone])

    max_width = max(map(len, tuning))
    tuning = [n.ljust(max_width) for n in tuning]
    return tuning


def render_song(
    song: guitarpro.Song,
    track_number: int = 0,
    config: Optional[RenderConfig] = None,
) -> str:
    if config is None:
        config = RenderConfig()

    tab = parse_song(song, track_number)
    measures = less_naive_render_beats(
        tab, n_strings=len(song.tracks[track_number].strings)
    )
    tuning = get_tuning(song.tracks[track_number].strings)
    body = render_measures(
        measures,
        line_length=config.line.line_length,
        show_lyrics=config.line.show_lyrics,
        bar_numbers=config.line.show_bar_numbers,
        tuning=tuning,
        lyrics_position=config.line.lyrics_position,
    )

    header_lines = []

    if config.header.show_title and song.title:
        if config.header.center_title:
            header_lines.append(song.title.center(config.line.line_length))
        else:
            header_lines.append(song.title)

        header_lines.append("")

    if config.header.show_subtitle and song.subtitle:
        if config.header.center_title:
            header_lines.append(song.subtitle.center(config.line.line_length))
        else:
            header_lines.append(song.subtitle)

        header_lines.append("")

    if config.header.show_album and song.album:
        header_lines.append(f"Album: {song.album}")

    if config.header.show_artist and song.artist:
        header_lines.append(f"Artist: {song.artist}")

    if config.header.show_music and song.music:
        header_lines.append(f"Music: {song.music}")

    if config.header.show_words and song.words:
        header_lines.append(f"Words: {song.words}")

    if config.header.show_copyright and song.copyright:
        header_lines.append(f"Copyright: {song.copyright}")

    if config.header.show_tab and song.tab:
        header_lines.append(f"Arrangement: {song.tab}")

    header = "\n".join(header_lines)

    output = io.StringIO()

    print(header, file=output)
    print(file=output)
    print(body, file=output)

    return strip_trailing_whitespace(output.getvalue())


@pytest.mark.parametrize(
    "sample",
    [
        "BeautyAndTheBeast.gp5",
        "DifferentNotes.gp5",
        "BasicSustain.gp5",
        "TieNote.gp5",
        "CarpetOfTheSun.gp5",
    ],
)
def test_parse_song(verify_tab, sample):
    with get_sample(sample).open("rb") as stream:
        song = guitarpro.parse(stream)

    tab = parse_song(song, 0)
    measures = less_naive_render_beats(tab, n_strings=len(song.tracks[0].strings))
    verify_tab(naive_render_measures(measures))


@pytest.mark.parametrize("line_length", [60, 80, 90])
@pytest.mark.parametrize(
    "sample",
    [
        "BeautyAndTheBeast.gp5",
        "DifferentNotes.gp5",
        "BasicSustain.gp5",
        "TieNote.gp5",
        "CarpetOfTheSun.gp5",
    ],
)
def test_render_song(verify_tab, sample, line_length):
    with get_sample(sample).open("rb") as stream:
        song = guitarpro.parse(stream)

    tab = parse_song(song, 0)
    measures = less_naive_render_beats(tab, n_strings=len(song.tracks[0].strings))
    verify_tab(render_measures(measures, line_length=line_length))


@pytest.mark.parametrize("line_length", [60, 80])
@pytest.mark.parametrize(
    "sample",
    [
        "BeautyAndTheBeast.gp5",
        "CarpetOfTheSun.gp5",
    ],
)
def test_render_full(verify_tab, sample, line_length):
    with get_sample(sample).open("rb") as stream:
        song = guitarpro.parse(stream)

    config = RenderConfig()
    config.line.lyrics_position = LyricsPosition.Top
    verify_tab(render_song(song, config=config))
