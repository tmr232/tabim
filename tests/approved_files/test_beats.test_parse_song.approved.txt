0
      Tale as old as

--------------0---1--
-----------1--====---
------2----===-------
---3--3----==========
---3--=====----------
1--1--1----==========

1
time,

======---
---------
3-----3--
5-----2--
3-----3--
======---

2
   true as it can

-----------0--1---
--------1--===----
===2----===-------
3--3----==========
3--=====----------
1--1----==========

3
be.

3---===
-------
3---3--
5---2--
3---3--
====---

4
      Bare ly e  ven

1--===1----3--5--6---
---------------------
===------------------
3--===---------------
===------------------
1--1--1----===3--====

5
friends, then some bod y

8--------8----6----5---3--
--------------------------
--------------------------
7--------=====------------
7--------=====-----3---===
5--------5----=========---

6
bends un ex pect ed

======6--5--=====---
6-----===---8----6--
--------------------
8-----===-----------
8-----===---5----===
6-----6--========---

7
ly.

-------
====---
5---3--
5---2--
3---3--
-------

8
   Just a  lit tle

-----------0---1---
--------1--====----
===2----===--------
3--3----===========
3--=====-----------
1--1----===========

9
change.

========---
-----------
3-------3--
5-------2--
3-------3--
========---

10
      Small, to say the

----------------0---1---
-------------1--====----
===---2------===--------
===3--3------===========
===3--=======-----------
1--===1------===========

11
least. Both a  lit tle

3------5----3--5---8---
-----------------------
3------=====-----------
=======3----===7---====
3------=====-----------
=======----------------

12
scared, nei ther one pre

========-------------5---
6-------6---5----6---====
-------------------------
========---------7---====
-------------------------
6-------6---=====6---====

13
pared.    Beau ty and the

=======---5----6--3---5---
--------------------------
3------========-----------
=======-------------------
----------3----===3---====
3------3--=====-----------

14
Beast.

1------===
----------
----------
----------
=======---
1------===

15
            Ev er just the

---------------3--5----6---
------------6--===---------
------3--===---------------
---5--===5--===------------
3--3--3--3--3--===3----====
===------------------------

16
same.

8-----======
---------10-
------9--===
7-----===---
======------
------------

17
   Ev er a  sur

===---------10--
===------6--====
---7--9--===----
----------------
----------------
6--6--===6--====

18
prise.

8------======
----------10-
-------9--===
7------===---
-------------
=======------

19
   E  ver as be

===----------6--
===6--8---10-===
----------------
----------------
----------------
6--6--====6--===

20
fore, ev er just as

8-----8--6--=====---
------------10---8--
--------------------
7-----7--========---
--------------------
======------8----===

21
sure as the sun will

------------5---=====
6----6--8---====6----
---------------------
0----0--====0---=====
---------------------
=====----------------

22
rise.

---------
8-----===
---------
======3--
6-----===
------1--

23
   Tale as old as

-----------2---3--
===-----3--====---
---4----===-------
5--5----==========
------------------
3--3----==========

24
time.

======---
---------
5-----5--
7-----4--
5-----5--
======---

25
   Tune as old as

-----------2---3--
--------3--====---
===4----===-------
5--5----==========
===---------------
3--3----==========

26
song.

5-----===
---------
5-----5--
7-----4--
5-----5--
======---

27
   Bit ter sweet and

3--3---5---7-----8---
---------------------
4--====--------------
===------------------
===------------------
3--3---====5-----====

28
strange, find ing you can

10-------10---8---7---5---
--------------------------
--------------------------
--------------------------
------------------5---====
7--------7----========----

29
change, learn ing you were

========8-----7---====-----
8-------======----10--8----
---------------------------
---------------------------
========----------7---=====
8-------8-----========-----

30
wrong.

----------
=======---
7------5--
7------4--
5------5--
----------

31
   Cer tain as the

------------2--3---
-------3----===----
4--4---=====-------
5--====------------
5--====------------
3--3---============

32
sun

====---
-------
5---5--
7---4--
5---5--
====---

33
   ris ing in the

-----------2--3---
-------3---===----
4--4---====-------
5--====-----------
5--====-----------
3--3---===========

34
East. Tale as old as

5-----=====-------10-
------12---10-12--===
5-----=====----------
7-----=====----------
5-----10---==========
======---------------

35
time, song as old as

======------------7--
8-----8----7--8---===
---------------------
---------------------
======---------------
8-----7----==========

36
rhyme. Beau ty and the

=======7----8--5---7---
-----------------------
5------=====-----------
-------0----===========
-----------------------
5------=====-----------

37
Beast.

3------===
----------
----------
=======---
----------
3------2--

38
   Tale as old as

---------------10-
---12---10-12--===
------------------
------------------
------------------
0--0----==========

39
time, song as old as

======------------7--
8-----8----7--8---===
---------------------
---------------------
---------------------
8-----7----==========

40
rhyme. Beau ty and the

=======7----8--5---7---
-----------------------
5------=====-----------
-------0----===========
0------=====-----------
=======----------------

41
Beast.

3------
3------
4------
5------
5------
3------

